usb: device_next: allow to pass maximum possible MPS value to drivers

jfischer-no opened the pull request:
Allow the stack to pass the maximum possible MPS value within alternate
settings.

jfischer-no committed:
usb: device_next: allow to pass maximum possible MPS value to drivers

The wMaxPacketSize (MPS) value of an endpoint may vary depending on the
interface's alternate settings. Device controller, which uses dedicated
memory for the endpoint FIFO, may end up in a locked state during
endpoint reconfiguration if the allocated memory is too small but the
higher memory window is still used. Although the driver could allocate a
new, larger memory window, the last used window would likely remain
unused, wasting a bit of dedicated RAM.

Allow the stack to pass the maximum possible MPS value within alternate
settings. For a controller with endpoint FIFOs and dedicated RAM, the
driver should use this value for the FIFO configuration.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc: add helper to enable/disable control endpoint

Add helper function to enable/disable control endpoint.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_dwc2: allocate FIFO memory based on maximum MPS value

To ensure that the alternate endpoint setting does not result in a
memory window that is too small being allocated and locked because a
higher FIFO is still in use, use the maximum MPS value for FIFO memory
allocation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/device_next/usbd_endpoint.h:
```diff
@@ -73,12 +73,14 @@ static inline bool usbd_ep_bm_is_set(const uint32_t *const ep_bm, const uint8_t
  */
 int usbd_ep_enable(const struct device *dev,
 		   const struct usb_ep_descriptor *const ed,
+		   const uint16_t m_mps,
```
This seems to be solving the issue for DWC2 driver if the SPRAM is large enough and thresholding is not needed. However, it leaves the driver with no clue e.g. how large the RxFIFO should be until it is already too late (i.e. DWC2 driver can no longer reconfigure RxFIFO).

I think to determine the RxFIFO size, the driver would have to know beforehand (even before the first bus reset) what is the maximum possible maximum packet size of all OUT endpoints in all configurations (and how many OUT endpoints there are).

Then there is also the "do we need thresholding" problem, i.e. will the suitable RxFIFO and all the TxFIFOs fit simultaneously in SPRAM, or do we need to allocate e.g. smaller TxFIFOs and enable TX thresholding.

Do you have any idea how to solve this problem on top of this PR?

tmon-nordic replied:
Yes. Explicit feedback wMaxPacketSize at Full-Speed is 3.

jfischer-no committed:
usb: device_next: calculate the maximum possible UDC memory usage

UDC controllers with internal endpoint FIFO memory may require
information about the maximum possible memory usage to configure the
FIFOs or verify that the memory is large enough.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc: organize the initialization arguments in a struct

Organize the initialization arguments in a structure to make it easily
extensible. Since some future arguments may not be known beforehand,
move the controller initialization after the stack initialization.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc: allow driver to use the memory usage calculated by stack

Drivers that depend on memory usage information should select the
Kconfig option UDC_DRIVER_REQUIRES_MEMORY_STATS.
To save memory or when interfaces do not have any endpoints, the user
may disable memory usage information calculation in the stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: udc_rpi_pico: check memory usage calculated by stack

Check that the memory usage calculated by the stack does not exceed the
available DPRAM space for the interface endpoints.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
